runtime/common/src/migrations/local_currency.rs

```diff
@@ -0,0 +1,444 @@
+			let n_pre: u32 = Decode::decode(&mut pre_state.as_slice())
+				.expect("pre_ugprade provides a valid state; qed");
+			let n = orml_asset_registry::Metadata::<T>::iter_keys().count() as u32;
+			assert_eq!(n_pre, n);
+
+			let local_currency_id: CurrencyId = Local::get().into();
+			let local_meta =
+				<orml_asset_registry::Pallet<T> as Inspect>::metadata(&local_currency_id)
+					.expect("Local asset was just registered; qed");
+			log::info!("{LOG_PREFIX} CheckAssetIntegrity: Local meta exists");
+
+			for variant in AssetList::get().into_iter() {
+				log::info!("{LOG_PREFIX} Checking asset {:?}", variant);
+				let variant_meta = <orml_asset_registry::Pallet<T> as Inspect>::metadata(&variant)
+					.expect("Asset variant is registered; qed");
+				assert_eq!(variant_meta.decimals, local_meta.decimals);
+				assert_eq!(
+					variant_meta.additional.local_representation,
+					Some(Local::get())
+				);
+			}
+
+			log::info!("{LOG_PREFIX} POST UPGRADE: Finished");
+
+			Ok(())
+		}
+	}
+
+	fn init_local_representation(
+		currency_id: CurrencyId,
+		check_list: Vec<CurrencyId>,
+		local_asset_id: LocalAssetId,
+	) -> Option<LocalAssetId> {
+		if check_list.iter().any(|c| c == &currency_id) {
+			log::info!(
+				"{LOG_PREFIX} Set local representation of asset variant {:?}",
+				currency_id
+			);
+			Some(local_asset_id)
+		} else {
+			log::info!(
+				"{LOG_PREFIX} Skipping setting local representation of asset variant {:?}",
+				currency_id
+			);
+			None
+		}
+	}
+}
+
+pub mod migrate_pool_currency {
+	use orml_traits::asset_registry::Inspect;
+
+	use super::*;
+
+	const LOG_PREFIX: &str = "MigratePoolCurrency";
+
+	pub struct Migration<T, TargetPoolId, CurrencyOut, CurrencyIn>(
+		sp_std::marker::PhantomData<(T, TargetPoolId, CurrencyOut, CurrencyIn)>,
+	);
+
+	impl<T, TargetPoolId, CurrencyOut, CurrencyIn> OnRuntimeUpgrade
+		for Migration<T, TargetPoolId, CurrencyOut, CurrencyIn>
+	where
+		T: pallet_pool_system::Config
+			+ orml_asset_registry::Config<CustomMetadata = CustomMetadata, AssetId = CurrencyId>,
+		TargetPoolId: Get<<T as pallet_pool_system::Config>::PoolId>,
+		CurrencyOut: Get<CurrencyId>,
+		CurrencyIn: Get<CurrencyId>,
+		<T as orml_asset_registry::Config>::AssetId: From<CurrencyId>,
+		<T as pallet_pool_system::Config>::CurrencyId: From<CurrencyId>,
+	{
+		fn on_runtime_upgrade() -> Weight {
+			let to = CurrencyIn::get();
+			let from = CurrencyOut::get();
+			let mut weight = T::DbWeight::get().reads(2);
+
+			if let Some(true) = check_local_coupling::<T>(from, to) {
+				pallet_pool_system::Pool::<T>::mutate(TargetPoolId::get(), |maybe_pool| {
+					if let Some(pool) = maybe_pool {
+						pool.currency = to.into();
+					}
+				});
+				weight.saturating_accrue(T::DbWeight::get().writes(1));
+				log::info!("{LOG_PREFIX} Migrated pool currency");
+			} else {
+				log::info!("{LOG_PREFIX} Skipping pool currency migration");
+			}
+
+			weight
+		}
```

Comment:
We need to error out here if there are outstanding invest orders. Redemptions are fine, as they can be collected in the new local representation.

Reply:
CR

Reply:
[70587ae](https://github.com/centrifuge/centrifuge-chain/pull/1713/commits/70587ae68ba0fa741c6dcaac6c6adb6df5e34230)